src/main/scala/optics/internal/focus/features/project/ProjectCapture.scala

```diff
@@ -0,0 +1,191 @@
+            val fieldName = for {
+              classSym <- caseType.tpe.classSymbol
+              // TODO: This /seems/ to be non-deterministic every so often...?
+              // Verify this is invariant
+              memberField <- classSym.caseFields.lift(fieldIdx)
+            } yield memberField.name
+
+            // TODO: Prove whether this can actually fail
+            val optic = fieldName match {
+              case Some(name) =>
+                caseType.tpe.asType match {
+                  case '[t] =>
+                    val lambdaExpr =
+                      '{ (x: t) => ${ new Rewrite(boundSym, Select.unique('{x}.asTerm, name))(expr).asExprOf[To] } }
+                    FocusImpl[t, To](lambdaExpr)
+                }
+              case _ => throw new RuntimeException(s"Expansion failed catastrophically, ${caseExpr.show}")
+            }
+            reifyCases(cases).map(caseType -> optic :: _)
+          }
+        case expr :: _ =>
+          Left(FocusError.UnexpectedCodeStructure(expr.show))
+      }
+    }
+
+    // We /must/ use the fully-qualified names, otherwise `.tpe` doesn't work properly
+    val get = '{ (optic: Lens[From, To]) => optic.get }
+    val getOption = '{ (optic: POptional[From, From, To, To]) => optic.getOption }
+
+    // Flip the setters to make it easier to transform
+    val replace = '{ (optic: POptional[From, From, To, To]) => Function.flip(optic.replace) }
+
+    // Everything to turn our `caseOptic` into a fully-fledged optic
+    val optionalReify = ('{ Optional.apply[From, To] }.asTerm, getOption.asTerm, replace.asTerm)
+    val lensReify = ('{ Lens.apply[From, To]}.asTerm, get.asTerm, replace.asTerm)
+
+    // Each of the branches can theoretically be a different optic
+    val result = reifyCases(caseBranches).map { caseOptics =>
+
+      // TODO: This needs fixing once we have variance!
+      // Find the least powerful optic
+      val opticReify = caseOptics.map(_._2.asTerm.tpe.asType).foldLeft(lensReify) {
+        case (pair, '[Optional[u, v]])  => optionalReify
+        case (pair, '[Lens[u, v]]) if !Set(optionalReify).contains(pair) => lensReify
+        case (pair, _) => pair
+      }
+
+      val (cons, first, second) = opticReify
+
+      val setterExpr = reifyCPSCase[From](caseOptics, second)
+      val getterExpr = reifyCPSCase[From](caseOptics, first)
+
+      cons.tpe.asType match {
+        case tpe @ '[(t, u => v => w) => o] =>
+          '{
+             val makeOptic = ${ cons.asExprOf[tpe.Underlying] }
+             val getter = ${ getterExpr.asExprOf[t] }
+             val setter = ${ setterExpr.asExprOf[v => u => w] }
+             makeOptic(getter, Function.flip(setter))
+           }
+      }
+    }
+
+    result match {
+      case Right(expr) =>
+        expr.asTerm.tpe.asType match {
+          case '[Lens[From, To]] =>
+            '{ $self.embed[From, To, Lens[From, To]](EmbeddedOptic(${expr.asTerm.asExprOf[Lens[From, To]]})) }
```

Comment:
This expansion here is for later when the `EmbededOptic` becomes a witness; with inheritance this will be massively simplified.